### PR TITLE
[Blob] Fixed the ability to upload from a generator

### PR DESCRIPTION
**Pull request by tasherif-msft**
This resolves https://github.com/Azure/azure-sdk-for-python/issues/17418

### CODE REVIEW

**Comment by varun-shahdadpuri** [comment]
Can we please expedite this PR? We have been facing issues while trying to upload block blob using a generator.

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads.py [hunk 1]
```diff
@@ -541,10 +541,14 @@ def read(self, size):
                     chunk = chunk.encode(self.encoding)
                 data += chunk
                 count += len(chunk)
+        # This means count < size and what's leftover will be returned in this call.
         except StopIteration:
-            pass
+            self.leftover = b""
 
         if count > size:
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads.py:548]:
```suggestion
        if count >= size:
```

File: sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads.py [hunk 2]
```diff
@@ -541,10 +541,14 @@ def read(self, size):
...
             self.leftover = data[size:]
 
+        if count == size:
+            self.leftover = b""
+
```
Comment (xiafu-msft) [inline sdk/storage/azure-storage-blob/azure/storage/blob/_shared/uploads.py:553]:
```suggestion
```

**Review by xiafu-msft — APPROVED** [review]

**Pull request merged**